torch/_inductor/codegen/multi_kernel.py

```diff
@@ -323,6 +342,39 @@ def benchmark_sub_kernels(kernel_calls):
             for kernel_call in kernel_calls
         ]
 
+    # record_choice and lookup_choice are helper functions for cpp-wrapper
+    # codegen. The first pass use record_choice to keep the choice and
+    # the second pass do lookup by calling lookup_choice.
+    #
+    # An alternative that reused the multi-kernel cache does not work well
+    # since during codegen of the second pass, it's very hard to know the
+    # path for the cache file. Also reading the cache file need do some IO
+    # which can be slower.
+    @staticmethod
+    def record_choice(multi_kernel_name, choice):
+        """
+        Record the multi-kernel choice for cpp-wrapper first pass codegen
+        for the second pass.
+
+        We should do nothing if this function is not called during codegen.
+        """
+        from torch._inductor.graph import GraphLowering
+
+        if not isinstance(V.graph, GraphLowering):
+            return
+
+        if not V.graph.record_multi_kernel_choice:
+            return
+
+        V.graph.multi_kernel_to_choice[multi_kernel_name] = choice
```

Comment:
I believe the second run is a different instance of `V.graph`.  Are we caching the multi-kernel choice to disk?

If not, we should to reduce compile times.

Reply:
i think we are using the same graph object for the 2 passes of cpp-wrapper codegen: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L1065-L1106 . But the two passes use different scheduler and wrapper objects. The comment said scheduler maybe can be reused though.

For on disk cache, yea, we do have that to reduce compilation time. The reason we don't use that cache here is that:
1. it's hard to look up that cache in `MultiKernel.run_kernel` since we need figure out the path for the cache file which is decided by the content of the python shim for the multi-kernel.
2. look up the on-disk cache is slower than just lookup an in-memory cache.